driver-sync/src/test/functional/com/mongodb/client/ClientSideEncryptionCorpusTest.java

```diff
@@ -0,0 +1,292 @@
+            byte[] localKeyId = Base64.getDecoder().decode("LOCALAAAAAAAAAAAAAAAAA==");
+
+            if (method.equals("auto")) {
+                corpusCopied.append(field, corpus.get(field));
+                continue;
+            }
+
+            if (!method.equals("explicit")) {
+                throw new UnsupportedOperationException("Unsupported method: " + method);
+            }
+
+            String fullAlgorithmName = "AEAD_AES_256_CBC_HMAC_SHA_512-";
+            if (abbreviatedAlgorithName.equals("rand")) {
+                fullAlgorithmName += "Random";
+            } else if (abbreviatedAlgorithName.equals("det")) {
+                fullAlgorithmName += "Deterministic";
+            } else {
+                throw new UnsupportedOperationException("Unsupported algorithm: " + abbreviatedAlgorithName);
+            }
+
+            EncryptOptions opts = new EncryptOptions(fullAlgorithmName);
+            if (identifier.equals("id")) {
+                if (kms.equals("aws")) {
+                    opts.keyId(new BsonBinary(BsonBinarySubType.UUID_STANDARD, awsKeyId));
+                } else if (kms.equals("local")) {
+                    opts.keyId(new BsonBinary(BsonBinarySubType.UUID_STANDARD, localKeyId));
+                } else {
+                    throw new UnsupportedOperationException("Unsupported provider: " + kms);
+                }
+            } else if (identifier.equals("altname")) {
+                opts.keyAltName(kms);
+            } else {
+                throw new UnsupportedOperationException("Unsupported identifier: " + identifier);
+            }
+
+            try {
+                BsonValue encryptedValue = clientEncryption.encrypt(value, opts);
+                fieldDocument.put("value", encryptedValue);
+                corpusCopied.append(field, fieldDocument);
+            } catch (MongoException e) {
+                if (allowed) {
+                    throw e;
+                }
+                corpusCopied.append(field, fieldDocument);
+            }
+        }
+
+        // Step 6: insert corpusCopied
+        MongoCollection<BsonDocument> encryptedCollection = autoEncryptingClient.getDatabase("db")
+                .getCollection("coll", BsonDocument.class);
+        encryptedCollection.insertOne(corpusCopied);
+
+        // Step 7: check the auto decrypted document
+        BsonDocument corpusDecrypted = encryptedCollection.find(new BsonDocument()).first();
+        assertEquals(corpus, corpusDecrypted);
+
+        // Step 8: check the document with an unencrypted client
+        MongoCollection coll = client.getDatabase("db").getCollection("coll", BsonDocument.class);
+        BsonDocument corpusEncryptedActual = (BsonDocument) coll.find(new BsonDocument()).first();
+        BsonDocument corpusEncryptedExpected = bsonDocumentFromPath("corpus-encrypted.json");
+
+        for (String field : corpusEncryptedActual.keySet()) {
```

Comment:
Also updated in https://github.com/mongodb/specifications/pull/604, I think we should iterate over corpusEncryptedExpected instead. If the test was written incorrectly, and corpusEncryptedActual ended up omitting fields, we'd skip asserting and not detect that.